### PR TITLE
[WIP] Implement Guzzle 6 adapter

### PR DESCRIPTION
**Pull request by ddeboer**

### CODE REVIEW

**Comment by sagikazarmark** [comment]
Thanks. I will take a look at it in the evening.

Don't care about the failing tests: they are because diactoros require PHP 5.5 and we still test again 5.4. If we can decouple from a specific message implementation then this issue will go away.

File: src/Guzzle6HttpAdapter.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
+<?php
+
+namespace Http\Adapter;
+
+use FOS\HttpCache\Exception\ProxyUnreachableException;
```
Comment (sagikazarmark) [inline src/Guzzle6HttpAdapter.php]: Is that really a Friends of Symfony class in Guzzle 6?

File: src/Guzzle6HttpAdapter.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function sendRequest(RequestInterface $request)
+    {
+        try {
+            return $this->client->send($request);
+        } catch (TransferException $e) {
+            throw ProxyUnreachableException::proxyUnreachable($request, $e);
```
Comment (sagikazarmark) [inline src/Guzzle6HttpAdapter.php]: I am not sure I understand why you throw this exception here. `sendRequest` should throw an `HttpAdapterException`.

File: src/Guzzle6HttpAdapter.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+        if (count($exceptions) > 0) {
+            throw new MultiHttpAdapterException($exceptions);
+        }
+
+        return $results;
+    }
+
+    /**
+     * Returns the name
```
Comment (sagikazarmark) [inline src/Guzzle6HttpAdapter.php]: {@inheritdoc} here please

File: src/Guzzle6HttpAdapter.php [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+use GuzzleHttp\Pool;
+use GuzzleHttp\Promise\Promise;
+use Http\Adapter\Common\Exception\HttpAdapterException;
+use Http\Adapter\Common\Exception\MultiHttpAdapterException;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+class Guzzle6HttpAdapter implements PsrHttpAdapter
+{
+    public function __construct(Client $client = null)
```
Comment (sagikazarmark) [inline src/Guzzle6HttpAdapter.php]: Please add docblock without a description but with the parameter.

File: src/Guzzle6HttpAdapter.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+use GuzzleHttp\Exception\RequestException;
+use GuzzleHttp\Exception\TransferException;
+use GuzzleHttp\Pool;
+use GuzzleHttp\Promise\Promise;
+use Http\Adapter\Common\Exception\HttpAdapterException;
+use Http\Adapter\Common\Exception\MultiHttpAdapterException;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+class Guzzle6HttpAdapter implements PsrHttpAdapter
```
Comment (sagikazarmark) [inline src/Guzzle6HttpAdapter.php]: Please add author information here

File: src/Guzzle6HttpAdapter.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
+<?php
+
```
Comment (sagikazarmark) [inline src/Guzzle6HttpAdapter.php]: Missing file header, check guzzle5-adapter

**Comment by sagikazarmark** [comment]
Please check the guzzle5-adapter package to see how integration test should be added.

File: composer.json [hunk 7]
```diff
@@ -6,13 +6,14 @@
...
-            "email": "geloen.eric@gmail.com"
+            "name": "David de Boer",
+            "email": "david@ddeboer.nl"
         }
     ],
     "require": {
         "php": ">=5.4",
-        "php-http/adapter-core": "dev-internal_separation"
+        "php-http/adapter-core": "dev-internal_separation",
+        "guzzlehttp/guzzle": "6.x-dev"
```
Comment (ddeboer) [inline composer.json]: This should be changed to `~6.0` when Guzzle 6 has been released.

File: src/Guzzle6HttpAdapter.php (hunk 6 above)
Comment (ddeboer) [inline src/Guzzle6HttpAdapter.php]: Added.

**Comment by ddeboer** [comment]
Sorry, accidentally committed some WIP stuff. Fixed that.

> Please check the guzzle5-adapter package to see how integration test should be added.

I only see an `HttpAdapterTest`; which test class should I use for testing an PSR HTTP adapter?

**Comment by sagikazarmark** [comment]
See [this](https://github.com/php-http/guzzle5-adapter/blob/master/tests/Guzzle5HttpAdapterTest.php) file. The two methods you have to implement: `testGetName` and `createHttpAdapter`.

AFAIK Guzzle 6 uses the same ring architecture, so you should probably create multiple tests to test with various handlers.

**Comment by ddeboer** [comment]
Thanks. I meant: that’s the `HttpAdapterTest`, but I only implemented the minimal `Psr7Adapter`, for which I can find no test.

**Comment by sagikazarmark** [comment]
Ah, I missed that, I thought you extended the `CoreHttpAdapter`. Adapters provided by us should extend that class so that it is both a PsrHttpAdapter and an HttpAdapter. However it Guzzle6 is going to be the first adapter which is PSR compatible. It is still undecided how we should proceed in that case: force CoreHttpAdapter or have another implementation which helps implementing HttpAdapter interface.

**Comment by ddeboer** [comment]
Yes, things becomes different in the case of Guzzle 6, which does PSR-7 natively. 

My intuition is that we don’t really need CoreHttpAdapter in this case. I like the minimalist PsrHttpAdapter interface, which just takes PSR-7 request objects. Now PSR-7 has been accepted, more and more users will probably start working with those objects in their projects. This makes the HttpAdapter (with `get()`/`post()` etc. convenience methods) less relevant here, as constructing a request object that is native to the adapter is no longer necessary. Do you think those convenience methods still add enough value to maintain them?

In any case, I do think we need a new ConfigurablePsrHttpAdapter to use `sendRequest(s)` with an `$option` argument.

**Comment by sagikazarmark** [comment]
I am not entirely sure that HttpAdapter should always be implemented.

However extending CoreHttpAdapter is extremely easy: replace sendRequest and sendRequests methods with their sendInternal* methods.

I am waiting for some feedback from @egeloen with the upcoming interface separation. After that we can discuss how we should proceed with PSR7 clients (since it is the first one).

**Comment by ddeboer** [comment]
But do we still need the distinction between internal and PSR-7 requests when building adapters for PSR-7 libraries?

Okay, I’ll wait with adding tests until the interface segregation has been finished. Where is the discussion taking place?

**Comment by sagikazarmark** [comment]
There isn't really a discussion, because @egeloen is disappeared. I invited you to the Gitter room.

Internal requests are not just there to provide a bridge between PSR7 and non-PSR7 clients. They are also used for special type bodies, like files, or arrays.

**Comment by sagikazarmark** [comment]
Guzzle 6 is released

**Comment by ddeboer** [comment]
Thanks, I had already changed the version constraint.

**Comment by sagikazarmark** [comment]
I've been thinking about this whole thing with PSR and not PSR. I think the right choice would be to have a Client implementation which relies on adapters. This way adapters can remain PSR wrappers and the client implements the callable methods AND most of the internal request things.

**Comment by ddeboer** [comment]
I’m not sure I understand. Do you mean you want to keep the internal/PSR request separation and the HttpAdapter’s convenience methods?

**Comment by sagikazarmark** [comment]
I mean I want to move them to a separate implementation (Client) and use adapters there. All the internal request and convenience methods would be part of the HTTP Client implementation. Adapters would only be PSR wrapper this way (basically accepting requests and returning responses).

**Comment by ddeboer** [comment]
I like it, as it keeps the adapters lean and pure PSR-7. I guess adapters for non PSR-7-compliant HTTP clients such as Guzzle 5 would then convert between the PSR request object as produced by Http\Client and the Guzzle 5 request object (and vice versa for the response).

**Comment by sagikazarmark** [comment]
Exactly.

**Comment by sagikazarmark** [comment]
@ddeboer Thanks for your contribution. I am going to merge this and finish tests.

**Pull request merged**